src/dlsym_default.c

```diff
@@ -335,6 +299,99 @@ if (default_symbol_index) {
       }
     }
   }
+  *tags_p = tags;
+  *default_symbol_index_p = default_symbol_index;
+  
+  if (default_symbol_index)
+    return tags.base_addr + tags.symtab[default_symbol_index].st_value;
+  else
+    return NULL;
+}
+
+// Given a pseudo-handle, symbol name, and addr, returns the address of the symbol
+// with the given name of a default version found by the search order of the given
+// handle which is either RTLD_DEFAULT or RTLD_NEXT.
+void *dlsym_default_internal_flag_handler(void* handle, const char* symbol, void* addr,
+      dt_tag *tags_p, Elf32_Word *default_symbol_index_p)
+{
+  Dl_info info;
+  void* extra_info;
+  void* result;
+
+  // Retrieve the link_map for the library given by addr
+  int ret = dladdr1(addr, &info, &extra_info, RTLD_DL_LINKMAP);
+  if (!ret) {
+    printf("dladdr1 could not find shared object for address\n");
+    return NULL;
+  }
+
+  struct link_map* map = (struct link_map*) extra_info;
+
+  // Handle RTLD_DEFAULT starts search at first loaded object
+  if (handle == RTLD_DEFAULT) {
+    while (map->l_prev) {
+      // Rewinding to search by load order
+      map = map->l_prev;
+    }
+  }	
+
+  // Handle RTLD_NEXT starts search after current library
+  if (handle == RTLD_NEXT) {
+    // Skip current library
+    if (!map->l_next) {
+      printf("There are no libraries after the current library.\n");
+      return NULL;
+    }
+    map = map->l_next;
+  }
+
+  // Search through libraries until end of list is reached or symbol is found.
+  while (1) {
+    // printf("l_name: %s\n", map->l_name);
+    // Running dlsym_default_internal on the linux-vdso library will
+    // cause a segfault. Because it may have different versions on
+    // different systems, we check this way rather than a strcmp.
+    // If the library is vdso, we simply move to the next loaded,
+    // library if it exists.
+    if (map->l_name[0] == 'l' &&
+      map->l_name[1] == 'i' &&
+      map->l_name[2] == 'n' &&
+      map->l_name[3] == 'u' &&
+      map->l_name[4] == 'x' &&
+      map->l_name[5] == '-' &&
+      map->l_name[6] == 'v' &&
+      map->l_name[7] == 'd' &&
+      map->l_name[8] == 's' &&
+      map->l_name[9] == 'o') {
+			if (!map->l_next) {
```

Comment:
Nit: Fix indentation. I believe we use a 2-space indentation in this file. Let's stick to that.